File: lang/c/src/vtfppc/ImageConversion.cpp
#include <vtfppc/ImageConversion.h>

#include <vtfpp/ImageConversion.h>

#include <sourceppc/Convert.hpp>
#include <sourceppc/Helpers.h>

using namespace vtfpp;

SOURCEPP_API sourcepp_buffer_t vtfpp_image_conversion_convert_image_data_to_format(const unsigned char* buffer, size_t bufferLen, vtfpp_image_format_e oldFormat, vtfpp_image_format_e newFormat, uint16_t width, uint16_t height) {
	SOURCEPP_EARLY_RETURN_VAL(buffer, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(bufferLen, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(oldFormat != VTFPP_IMAGE_FORMAT_EMPTY, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(newFormat != VTFPP_IMAGE_FORMAT_EMPTY, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(width, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(height, SOURCEPP_BUFFER_INVALID);

	return Convert::toBuffer(ImageConversion::convertImageDataToFormat({reinterpret_cast<const std::byte*>(buffer), bufferLen}, static_cast<ImageFormat>(oldFormat), static_cast<ImageFormat>(newFormat), width, height));
}

SOURCEPP_API sourcepp_buffer_t vtfpp_convert_several_image_data_to_format(const unsigned char* buffer, size_t bufferLen, vtfpp_image_format_e oldFormat, vtfpp_image_format_e newFormat, uint8_t mipCount, uint16_t frameCount, uint8_t faceCount, uint16_t width, uint16_t height, uint16_t sliceCount) {
	SOURCEPP_EARLY_RETURN_VAL(buffer, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(bufferLen, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(oldFormat != VTFPP_IMAGE_FORMAT_EMPTY, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(newFormat != VTFPP_IMAGE_FORMAT_EMPTY, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(mipCount, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(frameCount, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(faceCount == 0 || faceCount == 6 || faceCount == 7, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(width, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(height, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(sliceCount, SOURCEPP_BUFFER_INVALID);

	return Convert::toBuffer(ImageConversion::convertSeveralImageDataToFormat({reinterpret_cast<const std::byte*>(buffer), bufferLen}, static_cast<ImageFormat>(oldFormat), static_cast<ImageFormat>(newFormat), mipCount, frameCount, faceCount, width, height, sliceCount));
}

SOURCEPP_API sourcepp_buffer_t vtfpp_image_conversion_convert_hdri_to_cubemap(const unsigned char* buffer, size_t bufferLen, vtfpp_image_format_e format, uint16_t width, uint16_t height) {
	SOURCEPP_EARLY_RETURN_VAL(buffer, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(bufferLen, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(format != VTFPP_IMAGE_FORMAT_EMPTY, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(width, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(height, SOURCEPP_BUFFER_INVALID);

	const auto faces = ImageConversion::convertHDRIToCubeMap({reinterpret_cast<const std::byte*>(buffer), bufferLen}, static_cast<ImageFormat>(format), width, height);
	std::vector<std::byte> out;
	for (int i = 0; i < faces.size(); i++) {
		out.insert(out.end(), faces[i].begin(), faces[i].end());
	}
	return Convert::toBuffer(out);
}

SOURCEPP_API sourcepp_buffer_t vtfpp_image_conversion_convert_hdri_to_cubemap_ex(const unsigned char* buffer, size_t bufferLen, vtfpp_image_format_e format, uint16_t width, uint16_t height, uint16_t resolution, int bilinear) {
	SOURCEPP_EARLY_RETURN_VAL(buffer, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(bufferLen, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(format != VTFPP_IMAGE_FORMAT_EMPTY, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(width, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(height, SOURCEPP_BUFFER_INVALID);

	const auto faces = ImageConversion::convertHDRIToCubeMap({reinterpret_cast<const std::byte*>(buffer), bufferLen}, static_cast<ImageFormat>(format), width, height, resolution, bilinear);
	std::vector<std::byte> out;
	for (int i = 0; i < faces.size(); i++) {
		out.insert(out.end(), faces[i].begin(), faces[i].end());
	}
	return Convert::toBuffer(out);
}

SOURCEPP_API vtfpp_image_conversion_file_format_e vtfpp_image_conversion_get_default_file_format_for_image_format(vtfpp_image_format_e format) {
	return static_cast<vtfpp_image_conversion_file_format_e>(ImageConversion::getDefaultFileFormatForImageFormat(static_cast<ImageFormat>(format)));
}

SOURCEPP_API sourcepp_buffer_t vtfpp_image_conversion_convert_image_data_to_file(const unsigned char* buffer, size_t bufferLen, vtfpp_image_format_e format, uint16_t width, uint16_t height, vtfpp_image_conversion_file_format_e fileFormat) {
	SOURCEPP_EARLY_RETURN_VAL(buffer, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(bufferLen, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(format != VTFPP_IMAGE_FORMAT_EMPTY, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(width, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(height, SOURCEPP_BUFFER_INVALID);

	return Convert::toBuffer(ImageConversion::convertImageDataToFile({reinterpret_cast<const std::byte*>(buffer), bufferLen}, static_cast<ImageFormat>(format), width, height, static_cast<ImageConversion::FileFormat>(fileFormat)));
}

SOURCEPP_API sourcepp_buffer_t vtfpp_image_conversion_convert_file_to_image_data(const unsigned char* buffer, size_t bufferLen, vtfpp_image_format_e* format, int* width, int* height, int* frameCount) {
	SOURCEPP_EARLY_RETURN_VAL(buffer, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(bufferLen, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(format, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(width, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(height, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(frameCount, SOURCEPP_BUFFER_INVALID);

	return Convert::toBuffer(ImageConversion::convertFileToImageData({reinterpret_cast<const std::byte*>(buffer), bufferLen}, reinterpret_cast<ImageFormat&>(*format), *width, *height, *frameCount));
}

SOURCEPP_API uint16_t vtfpp_image_conversion_get_resized_dim(uint16_t n, vtfpp_image_conversion_resize_method_e method) {
	return ImageConversion::getResizedDim(n, static_cast<ImageConversion::ResizeMethod>(method));
}

SOURCEPP_API void vtfpp_image_conversion_set_resized_dims(uint16_t* width, vtfpp_image_conversion_resize_method_e widthResize, uint16_t* height, vtfpp_image_conversion_resize_method_e heightResize) {
	SOURCEPP_EARLY_RETURN(width);
	SOURCEPP_EARLY_RETURN(height);

	ImageConversion::setResizedDims(*width, static_cast<ImageConversion::ResizeMethod>(widthResize), *height, static_cast<ImageConversion::ResizeMethod>(heightResize));
}

SOURCEPP_API sourcepp_buffer_t vtfpp_image_conversion_resize_image_data(const unsigned char* buffer, size_t bufferLen, vtfpp_image_format_e format, uint16_t width, uint16_t newWidth, uint16_t height, uint16_t newHeight, int srgb, vtfpp_image_conversion_resize_filter_e filter, vtfpp_image_conversion_resize_edge_e edge) {
	SOURCEPP_EARLY_RETURN_VAL(buffer, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(bufferLen, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(format != VTFPP_IMAGE_FORMAT_EMPTY, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(width, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(newWidth, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(height, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(newHeight, SOURCEPP_BUFFER_INVALID);

	return Convert::toBuffer(ImageConversion::resizeImageData({reinterpret_cast<const std::byte*>(buffer), bufferLen}, static_cast<ImageFormat>(format), width, newWidth, height, newHeight, srgb, static_cast<ImageConversion::ResizeFilter>(filter), static_cast<ImageConversion::ResizeEdge>(edge)));
}

SOURCEPP_API sourcepp_buffer_t vtfpp_image_conversion_resize_image_data_strict(const unsigned char* buffer, size_t bufferLen, vtfpp_image_format_e format, uint16_t width, uint16_t newWidth, uint16_t* widthOut, vtfpp_image_conversion_resize_method_e widthResize, uint16_t height, uint16_t newHeight, uint16_t* heightOut, vtfpp_image_conversion_resize_method_e heightResize, int srgb, vtfpp_image_conversion_resize_filter_e filter, vtfpp_image_conversion_resize_edge_e edge) {
	SOURCEPP_EARLY_RETURN_VAL(buffer, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(bufferLen, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(format != VTFPP_IMAGE_FORMAT_EMPTY, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(width, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(newWidth, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(widthOut, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(height, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(newHeight, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(heightOut, SOURCEPP_BUFFER_INVALID);

	return Convert::toBuffer(ImageConversion::resizeImageDataStrict({reinterpret_cast<const std::byte*>(buffer), bufferLen}, static_cast<ImageFormat>(format), width, newWidth, *widthOut, static_cast<ImageConversion::ResizeMethod>(widthResize), height, newHeight, *heightOut, static_cast<ImageConversion::ResizeMethod>(heightResize), srgb, static_cast<ImageConversion::ResizeFilter>(filter), static_cast<ImageConversion::ResizeEdge>(edge)));
}

SOURCEPP_API sourcepp_buffer_t vtfpp_image_conversion_crop_image_data(const unsigned char* buffer, size_t bufferLen, vtfpp_image_format_e format, uint16_t width, uint16_t newWidth, uint16_t xOffset, uint16_t height, uint16_t newHeight, uint16_t yOffset) {
	SOURCEPP_EARLY_RETURN_VAL(buffer, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(bufferLen, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(format != VTFPP_IMAGE_FORMAT_EMPTY, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(width, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(newWidth, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(height, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(newHeight, SOURCEPP_BUFFER_INVALID);

	return Convert::toBuffer(ImageConversion::cropImageData({reinterpret_cast<const std::byte*>(buffer), bufferLen}, static_cast<ImageFormat>(format), width, newWidth, xOffset, height, newHeight, yOffset));
}

SOURCEPP_API sourcepp_buffer_t vtfpp_image_conversion_gamma_correct_image_data(const unsigned char* buffer, size_t bufferLen, vtfpp_image_format_e format, uint16_t width, uint16_t height, float gamma) {
	SOURCEPP_EARLY_RETURN_VAL(buffer, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(bufferLen, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(format != VTFPP_IMAGE_FORMAT_EMPTY, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(width, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(height, SOURCEPP_BUFFER_INVALID);

	return Convert::toBuffer(ImageConversion::gammaCorrectImageData({reinterpret_cast<const std::byte*>(buffer), bufferLen}, static_cast<ImageFormat>(format), width, height, gamma));
}

SOURCEPP_API sourcepp_buffer_t vtfpp_image_conversion_invert_green_channel_for_image_data(const unsigned char* buffer, size_t bufferLen, vtfpp_image_format_e format, uint16_t width, uint16_t height) {
	SOURCEPP_EARLY_RETURN_VAL(buffer, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(bufferLen, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(format != VTFPP_IMAGE_FORMAT_EMPTY, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(width, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(height, SOURCEPP_BUFFER_INVALID);

	return Convert::toBuffer(ImageConversion::invertGreenChannelForImageData({reinterpret_cast<const std::byte*>(buffer), bufferLen}, static_cast<ImageFormat>(format), width, height));
}

File: lang/csharp/src/vtfpp/ImageConversion.cs
using System;
using System.Runtime.InteropServices;

namespace sourcepp.vtfpp;

public static class ImageConversion
{
	public const float DEFAULT_COMPRESSED_QUALITY = -1;

	public static byte[] ConvertImageDataToFormat(ReadOnlySpan<byte> buffer, ImageFormat oldFormat, ImageFormat newFormat, ushort width, ushort height, float quality = DEFAULT_COMPRESSED_QUALITY)
	{
		return new sourcepp.Buffer(DLL.vtfpp_image_conversion_convert_image_data_to_format(buffer, (ulong) buffer.Length, oldFormat, newFormat, width, height, quality)).Read<byte>();
	}
	
	public static byte[] ConvertSeveralImageDataToFormat(ReadOnlySpan<byte> buffer, ImageFormat oldFormat, ImageFormat newFormat, byte mipCount, ushort frameCount, byte faceCount, ushort width, ushort height, ushort depth, float quality = DEFAULT_COMPRESSED_QUALITY)
	{
		return new sourcepp.Buffer(DLL.vtfpp_image_conversion_convert_several_image_data_to_format(buffer, (ulong) buffer.Length, oldFormat, newFormat, mipCount, frameCount, faceCount, width, height, depth, quality)).Read<byte>();
	}

	public static byte[] ConvertHdriToCubemap(ReadOnlySpan<byte> buffer, ImageFormat format, ushort width, ushort height, ushort resolution = 0, bool bilinear = true)
	{
		return new sourcepp.Buffer(DLL.vtfpp_image_conversion_convert_hdri_to_cubemap_ex(buffer, (ulong) buffer.Length, format, width, height, resolution, Convert.ToInt32(bilinear))).Read<byte>();
	}
	
	public enum FileFormat {
		DEFAULT = 0,
		PNG     = 1,
		JPG     = 2,
		BMP     = 3,
		TGA     = 4,
		WEBP    = 5,
		QOI     = 6,
		HDR     = 7,
		EXR     = 8,
	}

	public static FileFormat GetDefaultFileFormatForImageFormat(ImageFormat format)
	{
		return DLL.vtfpp_image_conversion_get_default_file_format_for_image_format(format);
	}

	public static byte[] ConvertImageDataToFile(ReadOnlySpan<byte> buffer, ImageFormat format, ushort width, ushort height, FileFormat fileFormat = FileFormat.DEFAULT)
	{
		return new sourcepp.Buffer(DLL.vtfpp_image_conversion_convert_image_data_to_file(buffer, (ulong) buffer.Length, format, width, height, fileFormat)).Read<byte>();
	}

	public static byte[] ConvertImageDataToFile(ReadOnlySpan<byte> buffer, ref ImageFormat format, ref int width, ref int height, ref int frameCount)
	{
		return new sourcepp.Buffer(DLL.vtfpp_image_conversion_convert_file_to_image_data(buffer, (ulong) buffer.Length, ref format, ref width, ref height, ref frameCount)).Read<byte>();
	}

	public enum ResizeEdge {
		CLAMP = 0,
		REFLECT,
		WRAP,
		ZERO,
	}

	public enum ResizeFilter {
		DEFAULT = 0,
		BOX,
		BILINEAR,
		CUBIC_BSPLINE,
		CATMULL_ROM,
		MITCHELL,
		POINT_SAMPLE,
		KAISER = 100,
		NICE = 101,
	}

	public enum ResizeMethod {
		NONE,
		POWER_OF_TWO_BIGGER,
		POWER_OF_TWO_SMALLER,
		POWER_OF_TWO_NEAREST,
	}

	[StructLayout(LayoutKind.Sequential)]
	public struct ResizeBounds
	{
		public ushort ResizeMinWidth = 0;
		public ushort ResizeMaxWidth = 0;
		public ushort ResizeMinHeight = 0;
		public ushort ResizeMaxHeight = 0;

		public ResizeBounds()
		{
		}

		public ResizeBounds(ushort size)
		{
			ResizeMinWidth = size;
			ResizeMaxWidth = size;
			ResizeMinHeight = size;
			ResizeMaxHeight = size;
		}

		public ResizeBounds(ushort width, ushort height)
		{
			ResizeMinWidth = width;
			ResizeMaxWidth = width;
			ResizeMinHeight = height;
			ResizeMaxHeight = height;
		}

		public ResizeBounds(ushort minWidth, ushort maxWidth, ushort minHeight, ushort maxHeight)
		{
			ResizeMinWidth = minWidth;
			ResizeMaxWidth = maxWidth;
			ResizeMinHeight = minHeight;
			ResizeMaxHeight = maxHeight;
		}
	}

	public static ushort GetResizedDim(ushort n, ResizeMethod method)
	{
		return DLL.vtfpp_image_conversion_get_resized_dim(n, method);
	}

	public static void SetResizedDims(ref ushort width, ResizeMethod widthResize, ref ushort height, ResizeMethod heightResize)
	{
		DLL.vtfpp_image_conversion_set_resized_dims(ref width, widthResize, ref height, heightResize);
	}

	public static byte[] ResizeImageData(ReadOnlySpan<byte> buffer, ImageFormat format, ushort width, ushort newWidth, ushort height, ushort newHeight, bool srgb, ResizeFilter filter, ResizeEdge edge = ResizeEdge.CLAMP)
	{
		return new sourcepp.Buffer(DLL.vtfpp_image_conversion_resize_image_data(buffer, (ulong) buffer.Length, format, width, newWidth, height, newHeight, Convert.ToInt32(srgb), filter, edge)).Read<byte>();
	}
	
	public static byte[] ResizeImageDataStrict(ReadOnlySpan<byte> buffer, ImageFormat format, ushort width, ushort newWidth, ref ushort widthOut, ResizeMethod widthResize, ushort height, ushort newHeight, ref ushort heightOut, ResizeMethod heightResize, bool srgb, ResizeFilter filter, ResizeEdge edge = ResizeEdge.CLAMP)
	{
		return new sourcepp.Buffer(DLL.vtfpp_image_conversion_resize_image_data_strict(buffer, (ulong) buffer.Length, format, width, newWidth, ref widthOut, widthResize, height, newHeight, ref heightOut, heightResize, Convert.ToInt32(srgb), filter, edge)).Read<byte>();
	}

	public static byte[] CropImageData(ReadOnlySpan<byte> buffer, ImageFormat format, ushort width, ushort newWidth, ushort xOffset, ushort height, ushort newHeight, ushort yOffset)
	{
		return new sourcepp.Buffer(DLL.vtfpp_image_conversion_crop_image_data(buffer, (ulong) buffer.Length, format, width, newWidth, xOffset, height, newHeight, yOffset)).Read<byte>();
	}

	public static void GammaCorrectImageData(Span<byte> buffer, ImageFormat format, ushort width, ushort height, float gamma)
	{
		DLL.vtfpp_image_conversion_gamma_correct_image_data(buffer, (ulong) buffer.Length, format, width, height, gamma);
	}

	public static void InvertGreenChannelForImageData(Span<byte> buffer, ImageFormat format, ushort width, ushort height)
	{
		DLL.vtfpp_image_conversion_invert_green_channel_for_image_data(buffer, (ulong) buffer.Length, format, width, height);
	}
}
